sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/coders/JAXBCoder.java

```diff
@@ -71,13 +73,25 @@ public void encode(T value, OutputStream outStream, Context context)
         JAXBContext jaxbContext = JAXBContext.newInstance(jaxbClass);
         jaxbMarshaller = jaxbContext.createMarshaller();
       }
-
-      jaxbMarshaller.marshal(value, new FilterOutputStream(outStream) {
-        // JAXB closes the underyling stream so we must filter out those calls.
-        @Override
-        public void close() throws IOException {
+      if (!context.isWholeStream) {
+        try {
+          long size = getEncodedElementByteSize(value, Context.OUTER);
+          // record the number of bytes the XML consists of so when reading we only read the encoded
+          // value
+          VarInt.encode(size, outStream);
+        } catch (Exception e) {
+          throw new CoderException(
+              "An Exception occured while trying to get the size of an encoded representation", e);
         }
-      });
+      }
+
+      jaxbMarshaller.marshal(
+          value,
+          new FilterOutputStream(outStream) {
+            // JAXB closes the underlying stream so we must filter out those calls.
+            @Override
+            public void close() throws IOException {}
```

Comment:
This one should be backported.

Also, why filter out CloseIgnoringInputStream but not CloseIgnoringOutputStream?